pkg/analyzer/log_test.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2023 The K8sGPT Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package analyzer
+
+import (
+	"context"
+	"sort"
+	"testing"
+
+	"github.com/k8sgpt-ai/k8sgpt/pkg/common"
+	"github.com/k8sgpt-ai/k8sgpt/pkg/kubernetes"
+	"github.com/stretchr/testify/require"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/fake"
+)
+
+func TestLogAnalyzer(t *testing.T) {
+	config := common.Analyzer{
+		Client: &kubernetes.Client{
+			Client: fake.NewSimpleClientset(
+				&v1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "Pod1",
+						Namespace: "default",
+						Labels: map[string]string{
+							"Name":      "Pod1",
+							"Namespace": "default",
+						},
+					},
+					Spec: v1.PodSpec{
+						Containers: []v1.Container{
+							{
+								Name: "test-container1",
+							},
+							{
+								Name: "test-container2",
+							},
+						},
+					},
+				},
+				&v1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "Pod2",
+						Namespace: "default",
+						Labels: map[string]string{
+							"Name":      "Pod1",
+							"Namespace": "default",
+						},
+					},
+				},
+				&v1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "Pod3",
+						Namespace: "test-namespace",
+						Labels: map[string]string{
+							"Name":      "Pod1",
+							"Namespace": "test-namespace",
+						},
+					},
+				},
+				&v1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "Pod4",
+						Namespace: "default",
+						Labels: map[string]string{
+							"Name":      "Pod4",
+							"Namespace": "default",
+						},
+					},
+					Spec: v1.PodSpec{
+						Containers: []v1.Container{
+							{
+								Name: "test-container3",
+							},
+						},
+					},
+				},
+			),
+		},
+		Context:   context.Background(),
+		Namespace: "default",
+	}
+
+	logAnalyzer := LogAnalyzer{}
+	results, err := logAnalyzer.Analyze(config)
+	require.NoError(t, err)
+
+	sort.Slice(results, func(i, j int) bool {
+		return results[i].Name < results[j].Name
+	})
+
+	expectations := []string{"default/Pod1/test-container1", "default/Pod1/test-container2", "default/Pod4/test-container3"}
+
+	for i, expectation := range expectations {
+		require.Equal(t, expectation, results[i].Name)
+
+		for _, failure := range results[i].Error {
+			require.Equal(t, "fake logs", failure.Text)
```

Comment:
how does these fake pods have error logs??

Reply:
My had, I should've marked this PR as draft. I've added a TODO note in the description to find a method to edit the logs of the fake containers.

Currently, the logs of these containers are automatically set to "fake logs" during instantiation. This is why I've marked "fake logs" as an erroneous log.

Reply:
@JuHyung-Son, I tried but couldn't find any method to modify the logs of these fake containers :(

Could you please share any ideas for creating fake logs?

Reply:
yeah.. i think there`s no way to do it.
i think it would be better to define fake logs or `results[i].Error` corresponding to each pod. think it as mock like `fake.NewSimpleClientset` create fake pods.

and remove 'fake logs' in this line. it's only for testing.
`errorPattern = regexp.MustCompile('(error|exception|fail|fake logs)')`

Reply:
If I remove `fake logs` from `errorPattern = regexp.MustCompile('(error|exception|fail|fake logs)')`, then there won't be anything in the results at all.

Reply:
yeah. i confused before. If we define `Error`, test is meaningless.
I'm not sure what to do about this situation. Maybe it's best to leave now.

@AlexsJones how do you think?

Reply:
I think this was always going to be a difficult one to test.  I am not sure, will need to think about it

Reply:
maybe we can try doing something like this and stream fake logs

```
import (
	"bytes"
	"context"
	"io"
	"testing"

	corev1 "k8s.io/api/core/v1"
	cgtesting "k8s.io/client-go/testing"
)
fp := FakePods{
		Fake: &FakeCoreV1{Fake: &cgtesting.Fake{}},
		ns:   "default",
	}
	req := fp.GetLogs("foo", &corev1.PodLogOptions{})
	body, err := req.Stream(context.Background())
	if err != nil {
		t.Fatal("Stream pod logs:", err)
	}
	var buf bytes.Buffer
	n, err := io.Copy(&buf, body)
	if err != nil {
		t.Fatal("Read pod logs:", err)
	}
	if n == 0 {
		t.Fatal("Empty log")
	}
	err = body.Close()
	if err != nil {
		t.Fatal("Close response body:", err)
	}
}```
an [example](https://github.com/kubernetes/kubernetes/pull/91485/files)

Reply:
That looks very helpful, thanks a lot @arbreezy! I'll update this PR soon.

Reply:
nice overview, let me investigate a little bit more before we concede :)

Reply:
Sure thing.

Reply:
@arbreezy, any updates on this?

Reply:
My approach ended up, being really hacky and to be fair not worth it. I think if we want to test the regex we can do it independetly of mocking logs in a pod.

Reply:
Agreed, we could always test regex separately. As a last resort, do you think that it would be a good idea to raise a feature request to allow passing a string to initialize the pod logs?

Reply:
Let's go with what we have for now?

Reply:
Yes, makes sense.
I was OOO sorry for the radio silence

Reply:
No problem @arbreezy